File: app/src/lib/series.ts
/**
 * The series-architecture layer: 98 candidates for Episodes 2-99, the framing segments, seasons, chronology and
 * connections.
 *
 * Authoritative sources (see content/06-series-architecture/README.md):
 *   - research fields come from the lead card (lead-format.ts), never from the planning file;
 *   - planning comes from data/stories.json and data/connections.json, which this app shows READ-ONLY.
 * Episode numbers are derived from seasons[].eps. The Python tools (tools/series.py) apply the same rules and
 * scripts/check-data.mjs checks that both agree.
 */
import fs from "node:fs/promises";
import path from "node:path";
import { CONTENT_DIR, DIRS, getData, type Data } from "./content";
import { leadFields, type Lead, type LeadFields } from "./lead-format";
import { listLeads } from "./leads";
import { FAMILY, ISSUE_TYPES, LADDER, SPOILER_ISSUES, ladderFor, mainMissing, premiseProblem, type Family, type Issue, type IssueCode, type Ladder } from "./series-rules";

export { FAMILY, ISSUE_TYPES, LADDER, ladderFor, mainMissing, premiseProblem };
export type { Family, Issue, IssueCode, Ladder };

export const SERIES_DIR = path.join(CONTENT_DIR, DIRS.series);
export const STORIES_FILE = `${DIRS.series}/data/stories.json`;
export const CONNECTIONS_FILE = `${DIRS.series}/data/connections.json`;

/** Server-side spoiler setting. When "hide", chronology, continuity and Episode 100 material never leave the server. */
export const spoilersHidden = () => (process.env.EXPLORER_SPOILERS ?? "").toLowerCase() === "hide";

/* ------------------------------------------------------------------ */
/* Types                                                               */
/* ------------------------------------------------------------------ */

export type Season = {
  id: string; title: string; episodes: number; question: string; assumption: string; contrasts: string; shift: string;
  prepares: string; ep1_effect: string; piece: string; eps: string[]; status: string;
};
export type Phase = { id: string; name: string; description: string; status: string };
export type Anchor = { rank: number; kind: string; requirement: string; audience_first: string; ep100_later: string; status: string; evidence: string };
export type Segment = { id: string; name: string; chron_order: number; content: string; source: string; status: string };
export type Connection = {
  id: string; from_id: string; to_id: string; type: string; family: Family; explanation: string; basis: string;
  status: "proposal" | "source-reported" | "source-checked"; first_reading: string; later_reading: string;
  required: string; spoiler: string;
};
export type CalendarConflict = { id: string; status: string; title: string; stories: string[]; detail: string; needs: string };

export type Story = {
  id: string; // ST-007
  leadId: string;
  title: string;
  premise: string;
  candidateStatus: string;
  episode: number;
  season: string;
  posInSeason: number;
  seasonLen: number;
  seasonPiece: string;
  chronPhase: string;
  chronConfidence: string;
  calendar: string;
  anchor?: Anchor;
  lead: Lead | null;
  fields: LeadFields | null;
  sq: string;
  primaryIdea: string;
  supportingIdeas: string[];
  territory: string;
  researchStatus: string;
  review: Lead["review"] | null;
  ladder: Ladder;
  connections: Connection[];
  conflicts: string[];
  issues: Issue[];
  mainMissing: string;
};

export type Series = {
  spoilersHidden: boolean;
  stories: Story[];
  byId: Map<string, Story>;
  byLead: Map<string, Story>;
  seasons: Season[];
  phases: Phase[];
  segments: Segment[];
  connections: Connection[];
  conflicts: CalendarConflict[];
  rejected: Record<string, string>;
  ideaDifferentiation: Record<string, string>;
  about: string;
};

export const FAMILY_LABEL: Record<Family, string> = {
  philosophical: "Philosophical relationship",
  chronological: "Chronological relationship",
  continuity: "Object or footage continuity",
};

export const EVIDENCE_LABEL: Record<Connection["status"], string> = {
  proposal: "Proposal (editorial or production plan)",
  "source-reported": "Reported in sources not yet opened",
  "source-checked": "Checked against opened sources",
};

/* ------------------------------------------------------------------ */
/* Derivations (mirror tools/series.py)                                */
/* ------------------------------------------------------------------ */

function issuesFor(st: Story, d: Data, conflicts: CalendarConflict[]): Issue[] {
  const out: Issue[] = [];
  const add = (code: IssueCode, detail: string) => out.push({ code, detail });
  const { lead, fields: f, ladder } = st;
  const pp = premiseProblem(st.premise, f?.situation ?? "");
  if (pp) add("premise", pp);
  if (st.candidateStatus === "direction" || st.candidateStatus === "unverified-lead")
    add("unverified-lead", `Candidate status is ${st.candidateStatus}. ${f?.unresolved ?? ""}`.trim());
  if (!lead || !f) {
    add("reference", `No lead card found for ${st.leadId}.`);
    return out;
  }
  if (f.cardStoryId && f.cardStoryId !== st.id) add("reference", `Lead card names ${f.cardStoryId}, but this story is ${st.id}.`);
  for (const i of [lead.leadIdeaId, ...lead.supportingIdeaIds]) {
    const idea = d.ideas.get(i);
    if (!idea) add("reference", `Idea ${i} does not exist in the idea bank.`);
    else if (idea.status !== "accepted for research") add("reference", `Idea ${i} is “${idea.status}”, not accepted for research.`);
  }
  if (lead.review.subject_identified !== "yes") add("subject", "No individual, family or group has been identified or contacted.");
  if (ladder.opened < ladder.total) add("sources", `${ladder.opened} of ${ladder.total} cited sources opened and checked.`);
  if (lead.review.claim_review !== "supported") add("claim", `Core claim review: ${lead.review.claim_review}.`);
  if (lead.review.filming_access !== "confirmed") add("access", `Filming access: ${lead.review.filming_access}. Contact made: ${f.contactMade || "not recorded"}.`);
  for (const x of st.connections) {
    if (x.status === "source-reported") add("continuity", `${x.id} rests on sources that have not been opened and checked.`);
    else if ((x.family === "continuity" || x.family === "chronological") && x.status === "proposal")
      add("continuity", `${x.id} is a proposal that depends on work not yet done: ${x.required || "not stated"}`);
  }
  for (const c of conflicts) if (c.stories.includes(st.id) && c.status !== "resolved") add("calendar", `${c.id}: ${c.title}`);
  if (!f.nextAction) add("next-action", "Section 19 has no next research action.");
  else if (["", "unassigned"].includes(f.nextOwner) || ["", "unassigned"].includes(f.nextBy))
    add("next-action", `Next action has no owner or date (owner: ${f.nextOwner || "none"}; by: ${f.nextBy || "none"}).`);
  for (const s of ladder.outOfOrder) add("order", `“${s}” is recorded while an earlier step is not met.`);
  return out;
}

/* ------------------------------------------------------------------ */
/* Loading                                                             */
/* ------------------------------------------------------------------ */

type RawStories = {
  about: string;
  rejected: Record<string, string>;
  seasons: Season[];
  phases: Phase[];
  calendar_conflicts?: CalendarConflict[];
  anchors: Record<string, Anchor>;
  idea_differentiation?: Record<string, string>;
  stories: Record<string, { story_id: string; lead_id: string; title: string; premise: string; candidate_status: string; chron_phase: string; chron_confidence: string; calendar: string }>;
  segments: Record<string, Omit<Segment, "id">>;
};

async function loadSeries(): Promise<Series> {
  const hide = spoilersHidden();
  const d = await getData();
  const raw = JSON.parse(await fs.readFile(path.join(SERIES_DIR, "data", "stories.json"), "utf8")) as RawStories;
  const rawCx = JSON.parse(await fs.readFile(path.join(SERIES_DIR, "data", "connections.json"), "utf8")) as Omit<Connection, "family">[];
  const leads = new Map((await listLeads()).map((l) => [l.id, l]));

  let connections: Connection[] = rawCx.map((x) => ({ ...x, family: FAMILY[x.type] ?? "philosophical" }));
  let conflicts = raw.calendar_conflicts ?? [];
  let segments: Segment[] = Object.entries(raw.segments).map(([id, s]) => ({ id, ...s })).sort((a, b) => a.chron_order - b.chron_order);
  if (hide) {
    // Only story-to-story philosophical pairings of low spoiler weight survive; nothing that touches a framing segment.
    connections = connections.filter((x) => x.family === "philosophical" && x.spoiler === "low" && !x.from_id.startsWith("SEG-") && !x.to_id.startsWith("SEG-"));
    conflicts = [];
    segments = segments.filter((s) => s.id === "SEG-E001").map((s) => ({ ...s, chron_order: 0, source: "", status: "" }));
  }

  const stories: Story[] = [];
  let ep = 2;
  for (const s of raw.seasons) {
    s.eps.forEach((leadId, i) => {
      const base = raw.stories[leadId];
      const lead = leads.get(leadId) ?? null;
      const fields = lead ? leadFields(lead) : null;
      const st: Story = {
        id: base.story_id,
        leadId,
        title: base.title,
        premise: base.premise,
        candidateStatus: base.candidate_status,
        episode: ep++,
        season: s.id,
        posInSeason: i + 1,
        seasonLen: s.eps.length,
        seasonPiece: s.piece,
        chronPhase: hide ? "" : base.chron_phase,
        chronConfidence: hide ? "" : base.chron_confidence,
        calendar: hide ? "" : base.calendar,
        anchor: hide ? undefined : raw.anchors[leadId],
        lead,
        fields,
        sq: lead?.shortlistQuestion ?? "",
        primaryIdea: lead?.leadIdeaId ?? "",
        supportingIdeas: lead?.supportingIdeaIds ?? [],
        territory: lead?.territory ?? "",
        researchStatus: lead?.researchStatus ?? "",
        review: lead?.review ?? null,
        ladder: ladderFor(lead, fields),
        connections: [],
        conflicts: [],
        issues: [],
        mainMissing: "",
      };
      st.connections = connections.filter((x) => x.from_id === st.id || x.to_id === st.id);
      st.conflicts = conflicts.filter((c) => c.stories.includes(st.id)).map((c) => c.id);
      stories.push(st);
    });
  }
  for (const st of stories) {
    st.issues = issuesFor(st, d, conflicts).filter((i) => !hide || !SPOILER_ISSUES.has(i.code));
    st.mainMissing = mainMissing(st);
  }

  return {
    spoilersHidden: hide,
    stories,
    byId: new Map(stories.map((s) => [s.id, s])),
    byLead: new Map(stories.map((s) => [s.leadId, s])),
    seasons: raw.seasons.map((s) => ({ ...s, ep1_effect: hide ? "" : s.ep1_effect })),
    phases: hide ? [] : raw.phases,
    segments,
    connections,
    conflicts,
    rejected: raw.rejected,
    ideaDifferentiation: raw.idea_differentiation ?? {},
    about: raw.about,
  };
}

const g = globalThis as unknown as { __cwhSeries?: { key: string; data: Promise<Series> } };
const MODULE_VERSION = Date.now();

export function getSeries(): Promise<Series> {
  const key = `${MODULE_VERSION}:${spoilersHidden()}`;
  if (!g.__cwhSeries || g.__cwhSeries.key !== key) {
    const data = loadSeries().catch((e) => { g.__cwhSeries = undefined; throw e; });
    g.__cwhSeries = { key, data };
  }
  return g.__cwhSeries.data;
}

/* ------------------------------------------------------------------ */
/* Helpers used by several pages                                       */
/* ------------------------------------------------------------------ */

export function storiesForIdea(s: Series, ideaId: string) {
  return {
    primary: s.stories.filter((x) => x.primaryIdea === ideaId),
    supporting: s.stories.filter((x) => x.supportingIdeas.includes(ideaId)),
  };
}

export function storiesForConcept(s: Series, d: Data, conceptId: string) {
  const ideaIds = new Set([...d.ideas.values()].filter((i) => i.primaryConcept === conceptId || i.supportingConcepts.includes(conceptId)).map((i) => i.id));
  return s.stories.filter((x) => ideaIds.has(x.primaryIdea) || x.supportingIdeas.some((i) => ideaIds.has(i)));
}

/** Counts for dashboards. Each is a separate question; none is a percentage. */
export function seriesCounts(s: Series) {
  const st = s.stories;
  return {
    cataloged: st.length,
    evidencedSituation: st.filter((x) => x.candidateStatus === "evidenced-situation").length,
    unverifiedLead: st.filter((x) => x.candidateStatus === "unverified-lead").length,
    sourcesOpened: st.reduce((n, x) => n + x.ladder.opened, 0),
    sourcesTotal: st.reduce((n, x) => n + x.ladder.total, 0),
    storiesWithAnOpenedSource: st.filter((x) => x.ladder.opened > 0).length,
    evidenceReviewed: st.filter((x) => x.review && x.review.claim_review !== "not reviewed").length,
    claimSupported: st.filter((x) => x.review?.claim_review === "supported").length,
    subjectsIdentified: st.filter((x) => x.review?.subject_identified === "yes").length,
    consentDocumented: st.filter((x) => x.review?.consent_status === "documented").length,
    accessConfirmed: st.filter((x) => x.review?.filming_access === "confirmed").length,
    productionReady: st.filter((x) => x.ladder.stage === LADDER.length).length,
  };
}

export function segmentLabel(id: string) {
  return id === "SEG-E001" ? "Episode 1" : id === "SEG-E100A" ? "Episode 100, Part A" : id === "SEG-E100B" ? "Episode 100, Part B" : id;
}

File: content/06-series-architecture/tools/series.py
# -*- coding: utf-8 -*-
"""Shared loader for the series-architecture layer.

Reads each entity from its one authoritative source and derives everything else:

    lead research fields   content/05-story-leads/SL-*.md        (the lead card; the app's lead editor writes it)
    story planning         content/06-series-architecture/data/stories.json
    connections            content/06-series-architecture/data/connections.json
    ideas                  content/03-idea-bank/ideas/*.md
    desk-research archive  content/06-series-architecture/data/leads.json   (frozen; history only)

Episode numbers, season positions and anchor flags are derived from `seasons[].eps` and `anchors` in stories.json.
They are never stored twice. The app's loader (app/src/lib/series.ts and lead-format.ts) mirrors these rules, and
app/scripts/check-data.mjs checks that both produce the same values.
"""
import json, os, re, collections

H = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))          # 06-series-architecture
D = os.path.join(H, "data")
CONTENT = os.path.dirname(H)
REPO = os.path.dirname(CONTENT)
LEADS_DIR = os.path.join(CONTENT, "05-story-leads")
IDEAS_DIR = os.path.join(CONTENT, "03-idea-bank", "ideas")

# --------------------------------------------------------------------------------------------------------------
# Lead cards
# --------------------------------------------------------------------------------------------------------------

LEAD_FILE_RE = re.compile(r"^(SL-SQ\d{2}-\d{3}).*\.md$")

# Optional review fields on a lead card. Absent means the conservative default; nothing is inferred from prose.
REVIEW_DEFAULTS = {
    "subject_identified": "no",          # no | yes
    "consent_status": "none",            # none | partial | documented
    "filming_access": "none",            # none | requested | confirmed
    "claim_review": "not reviewed",      # not reviewed | partly supported | supported | contradicted
}
REVIEW_VALUES = {
    "subject_identified": ["no", "yes"],
    "consent_status": ["none", "partial", "documented"],
    "filming_access": ["none", "requested", "confirmed"],
    "claim_review": ["not reviewed", "partly supported", "supported", "contradicted"],
}


def _read(p):
    return open(p, encoding="utf-8").read().replace("\r\n", "\n")


def parse_front_matter(text):
    m = re.match(r"^---\n(.*?)\n---\n?(.*)$", text, re.S)
    if not m:
        return {}, text
    fm = {}
    for line in m.group(1).split("\n"):
        mm = re.match(r"^([A-Za-z_][\w-]*):\s*(.*)$", line)
        if not mm:
            continue
        val = re.sub(r"\s+#.*$", "", mm.group(2)).strip()
        if val.startswith("[") and val.endswith("]"):
            val = [x.strip().strip("'\"") for x in val[1:-1].split(",") if x.strip()]
        else:
            val = val.strip("'\"")
        fm[mm.group(1)] = val
    return fm, m.group(2)


def split_sections(body):
    out, cur, buf = collections.OrderedDict(), None, []
    for line in body.split("\n"):
        m = re.match(r"^##\s+(\d+)\.\s+(.+)$", line)
        if m:
            if cur is not None:
                out[cur] = "\n".join(buf).strip()
            cur, buf = int(m.group(1)), []
        elif cur is not None:
            buf.append(line)
    if cur is not None:
        out[cur] = "\n".join(buf).strip()
    return out


def _paras(s):
    return [p.strip() for p in re.split(r"\n\s*\n", s or "") if p.strip()]


def _table(s):
    rows = []
    for line in (s or "").split("\n"):
        if not line.startswith("|") or re.match(r"^\|\s*-", line):
            continue
        rows.append([c.strip() for c in line.strip().strip("|").split(" | ")])
    return rows[1:] if rows else []          # drop the header row


def _line(s, label):
    m = re.search(r"\*\*%s:\*\*\s*(.*)" % re.escape(label), s or "")
    return m.group(1).strip() if m else ""


def source_review(notes):
    """'opened' only when the card says so explicitly. Anything else is a search result."""
    n = (notes or "").lower()
    if "not opened" in n or "search result only" in n:
        return "search-only"
    if re.search(r"opened and (read|checked)", n):
        return "opened"
    return "search-only"


def parse_lead_card(path):
    text = _read(path)
    fm, body = parse_front_matter(text)
    secs = split_sections(body)
    title = ""
    m = re.search(r"^#\s+.+?—\s*(.+)$", body, re.M)
    if m:
        title = m.group(1).strip()

    s2 = _paras(secs.get(2, ""))
    subject = s2[0] if s2 and not s2[0].startswith("-") else ""

    place = district = province = ""
    pl = _line(secs.get(3, ""), "Place")
    mm = re.match(r"^(.*)\s\(([^;()]*);\s*([^()]*?)\s+province\)$", pl)
    if mm:
        place, district, province = mm.group(1).strip(), mm.group(2).strip(), mm.group(3).strip()
    else:
        place = pl

    s5 = _paras(secs.get(5, ""))
    if s5 and s5[0].startswith("Every point below"):
        s5 = s5[1:]
    situation = "\n\n".join(s5)

    s9 = _paras(secs.get(9, ""))
    if s9 and s9[0].startswith("**Nothing below"):
        s9 = s9[1:]
    activity = "\n\n".join(s9)

    sources = []
    for r in _table(secs.get(18, "")):
        if len(r) < 5 or not re.match(r"^S\d+", r[0]):
            continue
        url = re.search(r"<([^>]+)>", r[2])
        sources.append(dict(ref=r[0], type=r[1], url=url.group(1) if url else r[2], accessed=r[3], notes=r[4],
                            review=source_review(r[4])))
    claims = []
    for r in _table(secs.get(11, "")):
        if len(r) < 5 or not re.match(r"^F\d+", r[0]):
            continue
        claims.append(dict(ref=r[0], claim=r[1], source=r[2], checked=r[3], how=r[4]))
    unresolved = ""
    for r in _table(secs.get(13, "")):
        if r and r[0] == "U1" and len(r) > 1:
            unresolved = r[1]
    s15 = secs.get(15, "")
    s17 = secs.get(17, "")
    ethics = s17.split("\n|")[0].strip() if s17 else ""
    s19 = secs.get(19, "")
    story_m = re.search(r"\*\*Series-architecture story ID:\*\*\s*`(ST-\d{3})`", secs.get(1, ""))

    review = {k: str(fm.get(k, v) or v) for k, v in REVIEW_DEFAULTS.items()}
    return dict(
        lead_id=fm.get("lead_id", ""), file=os.path.relpath(path, REPO).replace(os.sep, "/"), title=title,
        sq=fm.get("shortlist_question", ""), primary_idea=fm.get("lead_idea_id", ""),
        supporting_ideas=fm.get("supporting_idea_ids", []) or [], territory=fm.get("human_territory", ""),
        research_status=fm.get("research_status", "unverified"),
        screenplay_readiness=fm.get("screenplay_readiness", "not ready"),
        subject=subject, place=place, district=district, province=province, situation=situation,
        change=secs.get(6, ""), activity=activity, sources=sources, claims=claims, unresolved=unresolved,
        contact_made=_line(s15, "Contact made"), access=_line(s15, "Gatekeepers or permissions needed"),
        access_level=_line(s15, "Current access level"), ethics=ethics,
        next_action=_line(s19, "Action"), next_owner=_line(s19, "Owner"), next_by=_line(s19, "By (date)"),
        hold_reason=secs.get(21, ""), card_story_id=story_m.group(1) if story_m else "",
        sections=secs, **review)


def load_lead_cards():
    out = {}
    for f in sorted(os.listdir(LEADS_DIR)):
        m = LEAD_FILE_RE.match(f)
        if not m:
            continue
        c = parse_lead_card(os.path.join(LEADS_DIR, f))
        if c["lead_id"] == m.group(1):
            out[c["lead_id"]] = c
    return out


# --------------------------------------------------------------------------------------------------------------
# Idea bank
# --------------------------------------------------------------------------------------------------------------

def load_ideas():
    out = {}
    for f in sorted(os.listdir(IDEAS_DIR)):
        if not f.endswith(".md"):
            continue
        t = _read(os.path.join(IDEAS_DIR, f))
        fm, body = parse_front_matter(t)
        i = fm.get("id", "")

        def sec(name):
            m = re.search(r"^## %s\s*\n+(.*?)(?=\n## |\Z)" % re.escape(name), body, re.S | re.M)
            return m.group(1) if m else ""
        sc = fm.get("supporting_concepts", [])
        out[i] = dict(id=i, title=str(fm.get("title", "")).strip('"'), status=fm.get("status", ""),
                      primary_concept=fm.get("primary_concept", ""),
                      supporting_concepts=re.findall(r"C\d{3}", " ".join(sc) if isinstance(sc, list) else sc),
                      question=" ".join(sec("Open human question").split("\n\n")[0].split()),
                      limits=" ".join(sec("Limits and alternatives").split()),
                      core=" ".join(sec("Core idea").split("\n\n")[0].split()))
    return out


# --------------------------------------------------------------------------------------------------------------
# Readiness ladder: seven separate questions, never inferred from prose
# --------------------------------------------------------------------------------------------------------------

LADDER = [
    ("found", "Source found through search"),
    ("opened", "Source opened and checked"),
    ("claim", "Claim supported"),
    ("subject", "Subject identified"),
    ("consent", "Consent confirmed"),
    ("access", "Filming access confirmed"),
    ("ready", "Ready for production"),
]


def ladder(lead):
    src = lead["sources"] if lead else []
    opened = sum(1 for s in src if s["review"] == "opened")
    met = {
        "found": len(src) > 0,
        "opened": len(src) > 0 and opened == len(src),
        "claim": bool(lead) and lead["claim_review"] == "supported",
        "subject": bool(lead) and lead["subject_identified"] == "yes",
        "consent": bool(lead) and lead["consent_status"] == "documented",
        "access": bool(lead) and lead["filming_access"] == "confirmed",
    }
    met["ready"] = all(met.values()) and bool(lead) and lead["research_status"] == "verified" \
        and lead["screenplay_readiness"] == "ready for selection review"
    stage = 0
    for k, _ in LADDER:
        if not met[k]:
            break
        stage += 1
    out_of_order = [LADDER[i][1] for i in range(stage + 1, len(LADDER)) if met[LADDER[i][0]]]
    return dict(met=met, stage=stage, stage_label=LADDER[stage - 1][1] if stage else "No source recorded",
                opened=opened, total=len(src), out_of_order=out_of_order)


# --------------------------------------------------------------------------------------------------------------
# Connections
# --------------------------------------------------------------------------------------------------------------

FAMILY = {
    "philosophical contrast": "philosophical", "philosophical development": "philosophical",
    "shared human question": "philosophical",
    "chronological continuation": "chronological", "context revealed later": "chronological",
    "object continuity": "continuity", "revisited person or place": "continuity", "visual or sound echo": "continuity",
}
CX_STATUSES = ["proposal", "source-reported", "source-checked"]


# --------------------------------------------------------------------------------------------------------------
# Everything together
# --------------------------------------------------------------------------------------------------------------

def load():
    S = json.load(open(os.path.join(D, "stories.json"), encoding="utf-8"))
    CX = json.load(open(os.path.join(D, "connections.json"), encoding="utf-8"))
    cards = load_lead_cards()
    ideas = load_ideas()
    archive = {l["lead_id"]: l for l in json.load(open(os.path.join(D, "leads.json"), encoding="utf-8"))}

    anchors = S["anchors"]
    stories = collections.OrderedDict()
    ep = 2
    for s in S["seasons"]:
        for pos, lid in enumerate(s["eps"], 1):
            base = S["stories"][lid]
            lead = cards.get(lid)
            st = dict(base)
            st.update(episode=ep, season=s["id"], pos_in_season=pos, season_len=len(s["eps"]),
                      season_piece=s.get("piece", ""), anchor="yes" if lid in anchors else "no",
                      lead=lead)
            for k in ("title", "sq", "primary_idea", "supporting_ideas", "territory", "place", "district", "province",
                      "situation", "change", "activity", "unresolved", "access", "ethics", "sources", "subject",
                      "research_status", "subject_identified", "consent_status", "filming_access", "claim_review"):
                if k == "title":
                    continue                     # the working title belongs to the story
                st[k] = lead[k] if lead else ([] if k in ("supporting_ideas", "sources") else "")
            st["ladder"] = ladder(lead)
            stories[base["story_id"]] = st
            ep += 1
    for sid, st in stories.items():
        st["connections"] = [x for x in CX if sid in (x["from_id"], x["to_id"])]
    return dict(raw=S, CX=CX, cards=cards, ideas=ideas, archive=archive, stories=stories,
                seasons=S["seasons"], phases=S["phases"], anchors=anchors, segments=S["segments"],
                rejected=S["rejected"], conflicts=S.get("calendar_conflicts", []),
                lead_of={v["story_id"]: v["lead_id"] for v in stories.values()},
                sid_of={v["lead_id"]: v["story_id"] for v in stories.values()})


# --------------------------------------------------------------------------------------------------------------
# Needs attention: generated from records only
# --------------------------------------------------------------------------------------------------------------

ISSUE_TYPES = [
    ("premise", "Incomplete premise"),
    ("unverified-lead", "Core situation not yet evidenced"),
    ("subject", "Subject not identified"),
    ("sources", "Sources not opened"),
    ("claim", "Claim awaiting verification"),
    ("access", "Access unknown"),
    ("continuity", "Continuity claim not yet supported"),
    ("calendar", "Calendar conflict"),
    ("reference", "Broken idea or lead reference"),
    ("next-action", "Next action missing or unassigned"),
    ("order", "Status recorded out of order"),
]


def premise_problem(st):
    p = (st.get("premise") or "").strip()
    sit = (st.get("situation") or "").strip()
    if not p:
        return "No premise recorded."
    if not re.search(r"[.!?][\"'”’)]?$", p):
        return "Premise does not end with a full sentence."
    if sit and len(p) < len(sit) and sit.startswith(p):
        return "Premise is a mechanical cut of the situation text."
    if len(p.split()) > 60:
        return "Premise is longer than 60 words."
    return ""


def issues_for(st, data):
    out = []
    add = lambda code, detail: out.append(dict(code=code, detail=detail))
    lead, lad = st["lead"], st["ladder"]
    pp = premise_problem(st)
    if pp:
        add("premise", pp)
    if st.get("candidate_status") in ("direction", "unverified-lead"):
        add("unverified-lead", "Candidate status is `%s`. %s" % (st["candidate_status"], st.get("unresolved") or ""))
    if not lead:
        add("reference", "No lead card found for %s." % st["lead_id"])
        return out
    if lead["card_story_id"] and lead["card_story_id"] != st["story_id"]:
        add("reference", "Lead card names %s, but this story is %s." % (lead["card_story_id"], st["story_id"]))
    ideas = data["ideas"]
    for i in [lead["primary_idea"]] + list(lead["supporting_ideas"]):
        if i not in ideas:
            add("reference", "Idea %s does not exist in the idea bank." % i)
        elif ideas[i]["status"] != "accepted for research":
            add("reference", "Idea %s is `%s`, not accepted for research." % (i, ideas[i]["status"]))
    if lead["subject_identified"] != "yes":
        add("subject", "No individual, family or group has been identified or contacted.")
    if lad["opened"] < lad["total"]:
        add("sources", "%d of %d cited sources opened and checked." % (lad["opened"], lad["total"]))
    if lead["claim_review"] != "supported":
        add("claim", "Core claim review: %s." % lead["claim_review"])
    if lead["filming_access"] != "confirmed":
        add("access", "Filming access: %s. Contact made: %s." % (lead["filming_access"], lead["contact_made"] or "not recorded"))
    for x in st["connections"]:
        if x["status"] == "source-reported":
            add("continuity", "%s rests on sources that have not been opened and checked." % x["id"])
        elif FAMILY.get(x["type"]) in ("continuity", "chronological") and x["status"] == "proposal":
            add("continuity", "%s is a proposal that depends on work not yet done: %s" % (x["id"], x.get("required") or "not stated"))
    for c in data["conflicts"]:
        if st["story_id"] in c.get("stories", []) and c.get("status") != "resolved":
            add("calendar", "%s: %s" % (c["id"], c["title"]))
    if not lead["next_action"]:
        add("next-action", "Section 19 has no next research action.")
    elif lead["next_owner"] in ("", "unassigned") or lead["next_by"] in ("", "unassigned"):
        add("next-action", "Next action has no owner or date (owner: %s; by: %s)." % (lead["next_owner"] or "none", lead["next_by"] or "none"))
    for step in lad["out_of_order"]:
        add("order", "“%s” is recorded while an earlier step is not met." % step)
    return out


def main_missing(st):
    lad = st["ladder"]
    if st.get("candidate_status") in ("direction", "unverified-lead"):
        return "Confirm the core situation (unverified lead)"
    if lad["stage"] >= len(LADDER):
        return "Nothing recorded as missing"
    key, label = LADDER[lad["stage"]]
    if key == "opened":
        return "Open and check sources (%d of %d opened)" % (lad["opened"], lad["total"])
    return {"found": "Find a source", "claim": "Check the core claim", "subject": "Identify a subject",
            "consent": "Obtain and document consent", "access": "Confirm filming access",
            "ready": "Human selection decision"}[key]
